cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java

```diff
@@ -74,21 +97,76 @@
 
   private final RunnableTaskLauncher runnableTaskLauncher;
   private final BiConsumer<Boolean, String> stopper;
+
   private final AtomicInteger inflightRequests = new AtomicInteger(0);
+
+  /**
+   * Holds the total number of requests that have been executed by this handler that should count toward max allowed.
+   */
+  private final AtomicInteger totalRequestsWithUserCodeExecution = new AtomicInteger(0);
+
   private final String metadataServiceEndpoint;
 
+  private volatile RestartState restartState = RestartState.NOT_TIMED_OUT;
+
   public TaskWorkerHttpHandlerInternal(CConfiguration cConf, Consumer<String> stopper) {
+    int containerKillAfterExecution = cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION, 0);
     this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
     this.metadataServiceEndpoint = cConf.get(Constants.TaskWorker.METADATA_SERVICE_END_POINT);
     this.stopper = (terminate, className) -> {
-      if (cConf.getBoolean(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION) && terminate) {
-        if (className != null) {
+      synchronized (this) {
+        switch (restartState) {
+          case RESTART_NOW:
+            stopper.accept(className);
+            return;
+          case RESTART_IF_NO_INFLIGHT_REQUEST:
+            if (inflightRequests.compareAndSet(0, 1)) {
```

Comment:
nit: same as above. All references to inflightRequests are within sync block, does it need to be an atomic?

Reply:
see above reply.